### PR TITLE
Jwakefield/phase schedules

### PR DESCRIPTION
**Pull request by JoelWakefield**
- these changes add a rudimentary gantt chart, but a better one can be created once the data structure has been updated from immutable to a mutable one - this will be in the next update

### CODE REVIEW

File: ProjectManagementApp/Components/Pages/Phases/PhaseSchedules.razor [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        }
+    </MudGrid>
+}
+
+
+@code {
+    const int DAYS_DISPLAYED = 12;
+    
+    private IEnumerable<GanttItem> GanttItems { get; set; } = new List<GanttItem>();
+    private List<DateTime> Dates { get; set; } = new List<DateTime>();
```
Comment (jwthurmond) [inline ProjectManagementApp/Components/Pages/Phases/PhaseSchedules.razor:71]: maybe switch to DateOnly

https://learn.microsoft.com/en-us/dotnet/api/system.dateonly

File: ProjectManagementApp/Data/PhaseSchedule.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿namespace ProjectManagementApp.Data
+{
+	public class PhaseSchedule
+	{
+        public string Id { get; set; } = Guid.NewGuid().ToString();
+        public DateTime CreatedOn { get; set; } = DateTime.UtcNow;
+        public string PhaseId { get; set; }
+        public DateTime Start { get; set; }
```
Comment (jwthurmond) [inline ProjectManagementApp/Data/PhaseSchedule.cs:8]: If Start and End are only concerned with the Date and not time, you could consider switching to dateonly.

https://learn.microsoft.com/en-us/dotnet/api/system.dateonly

File: ProjectManagementApp/Services/PhaseScheduleService.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        private IPhaseService phaseService = phaseService;
+
+        public IEnumerable<PhaseSchedule> GetSchedules() => dbContext.PhaseSchedules;
+
+        /// <summary>
+        /// Returns all schedules in the form of a gantt chart item.
+        /// TODO: only check for items which would be displayed on the chart.
+        /// </summary>
+        /// <returns></returns>
+        public async Task<IEnumerable<GanttItem>> GetGanttItems()
```
Comment (jwthurmond) [inline ProjectManagementApp/Services/PhaseScheduleService.cs:27]: It seems like there's no await in this async method. I'd recommend updating the phaseService.GetAllPhases() method to async and having it return an IEnumerable instead of a list.

File: ProjectManagementApp/Services/PhaseScheduleService.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        {
+            var phases = phaseService.GetAllPhases().ToList();
+            var items = new List<GanttItem>();
+
+            foreach (var phase in phases)
+            {
+                var schedule = await GetScheduleAsync(phase.Id);
+
+                if (schedule != null)
+                    items.Add(new GanttItem(schedule, phase.Name));
```
Comment (jwthurmond) [inline ProjectManagementApp/Services/PhaseScheduleService.cs:37]: You're only using the Name value from the Phases, it would be better to create a function to just get the Phase Names?

File: ProjectManagementApp/Services/PhaseScheduleService.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            }
+
+            return items.AsEnumerable();
```
Comment (jwthurmond) [inline ProjectManagementApp/Services/PhaseScheduleService.cs:40]: a List is already Enumerable I don't think you need the "AsEnumberable()" here right?

File: ProjectManagementApp/Services/PhaseScheduleService.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        }
+
+        public async Task<PhaseSchedule?> GetScheduleAsync(string phaseId)
+        {
+            return await dbContext.PhaseSchedules
+                .OrderByDescending(s => s.CreatedOn)
+                .FirstOrDefaultAsync(s => s.PhaseId == phaseId);
+        }
+
+        public async Task CreateScheduleAsync(PhaseSchedule schedule)
```
Comment (jwthurmond) [inline ProjectManagementApp/Services/PhaseScheduleService.cs:50]: Instead of passing around the data entities, you should have an CreatePhaseScheduleVm that you convert to the data object.

File: ProjectManagementApp/Services/PhaseScheduleService.cs (hunk 3 above)
Comment (JoelWakefield) [reply ProjectManagementApp/Services/PhaseScheduleService.cs:27]: There's an await for the `var schedule = await GetScheduleAsync(phase.Id);`

**Review by jwthurmond — APPROVED** [review]

File: ProjectManagementApp/Services/PhaseScheduleService.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        /// </summary>
+        /// <returns></returns>
+        public async Task<IEnumerable<GanttItem>> GetGanttItems()
+        {
+            var phases = phaseService.GetAllPhases().ToList();
+            var items = new List<GanttItem>();
+
+            foreach (var phase in phases)
+            {
+                var schedule = await GetScheduleAsync(phase.Id);
```
Comment (jwthurmond) [inline ProjectManagementApp/Services/PhaseScheduleService.cs:34]: if you can avoid doing another call within a foreach, it would end up working better, but it can also be optimized in a different PR

File: ProjectManagementApp/Services/PhaseScheduleService.cs (hunk 4 above)
Comment (JoelWakefield) [reply ProjectManagementApp/Services/PhaseScheduleService.cs:37]: we're also using the `phase.Id`, a few lines above.

File: ProjectManagementApp/Services/PhaseScheduleService.cs (hunk 7 above)
Comment (JoelWakefield) [reply ProjectManagementApp/Services/PhaseScheduleService.cs:34]: I plan on changing that next PR

File: ProjectManagementApp/Services/PhaseScheduleService.cs (hunk 6 above)
Comment (JoelWakefield) [reply ProjectManagementApp/Services/PhaseScheduleService.cs]: I've updated the method to use a view model
